app/models/end_user.rb

```diff
@@ -5,7 +5,7 @@ class EndUser < User
   ).map(&:to_sym).freeze
 
   def visible_courses
-    self.courses.where.not('roles.type': 'Student')
-        .or(self.courses.where('courses.is_hidden': false, 'roles.hidden': false))
+    self.courses.where('roles.hidden': false)
```

Comment:
This is good but we need to go further. If the course is hidden, a student should not be able to access it at all even through the url directly.

Probably the best way to do this is to add another pre_check function in application_policy.rb that does:

```
    deny! if role.student? and role.course.is_hidden
``` 

And then skip that pre check everywhere we're currently skipping the `role_exists?` pre check. 

I guess you could also just add that line to the `role_exists?` precheck function and that would work but then maybe I'd change the name of the function from `role_exists?` to something that better describes the updated function